### ISSUE TITLE
DC motor wastegate position table feed forward open loop table

### ISSUE THREAD

**Issue by rusefillc** [rusefi/rusefi#7150]
4x4 1 byte S08 elements

**Comment by ElDominio** [comment]
any progress on htis?

**Comment by kifir23917** [comment]
@rusefillc, do we need increase this ticket priority?

**Comment by ElDominio** [comment]
a comment was deleted is there anything on this?

**Comment by ElDominio** [comment]
Tune at: https://rusefi.com/online/index.php?vehicleName=NegroWastegate&user_id=675&sorts[uploadedMsq]=-1&sorts[uploadedLog]=-1

Lua is inside tune, but important bit is:

wgPid = Pid.new(.01, 0.1, 0.005, -70, 70)
wgPid : setOffset(0.1)
wgPid : get(10, 10)
wgPid : reset()

startPwm(1, 500, 0.1) --pwm1 is pwm
startPwm(2, 100, 0) --pwm2 is dir
startPwm(3, 1, 0) --pwm3 is dis

wgPos = getSensor("WastegatePosition")
		targetWgPos = table3d(1, RPMread / 4, mapPres)
		-- p, i, d, min, max

		wgPidVal = wgPid : get(targetWgPos, wgPos)
		
		if wgPidVal < 0 then
			setPwmDuty(2, 0)
			wgPidVal = -wgPidVal
		else
			setPwmDuty(2, 1)
		end
		
		setPwmDuty(1, wgPidVal/10)
		wgPid : reset()

**Comment by FDSoftware** [comment]
> 4x4 1 byte S08 elements
> feed forward open loop wastegate position table

@ElDominio I think the interface in TS needs to be made a little more intuitive; but currently, as soon as ETB-Wastegate is selected, the same table is used as feedforward as the PWM output, so there's no need to create a new one just for ETB.



Another additional thing is that we can now select both a PWM and ETB output at the same time to control the same Turbo, and I don't think that's a good idea!
See the images below:
![Image](https://github.com/user-attachments/assets/c50548a9-3c6a-41b3-bd0d-4a15a2b3b82f)

![Image](https://github.com/user-attachments/assets/96abf685-a28e-4d8e-b52d-839a8b4e8e21)

I propose changing the `enable` option from a bool to an enum so that you can select:
* disabled
* enabled (PWM)
* enabled (E-Waste Gate)

Then with this, disable the PWM output both in TS and in the firmware where appropriate, and add a helper text on the boost control panel so that the user knows that they have to configure the e-waste gate h-bridge from the e-waste gate actuator tab.

**Comment by rusefillc** [comment]
@FDSoftware yes, great idea! let's reuse existing PWM open loop table for DC motor feedforward!

@FDSoftware note that I've created smaller #7150 for the purpose of focusing on just the feedforward, separating that urgent effort from less urgent wider #7109

**Comment by rusefillc** [comment]
first pair

![Image](https://github.com/user-attachments/assets/27e0b98b-593e-4c70-9fb1-58eaf8f87547)

**Comment by FDSoftware** [comment]
> that's similar to how for ETB idle we have two feedforwards and two PIDs - one for desired IAC position, and another for PWM to hold that desired powition

Could it be that in both cases the second PID is the closed loop one?

![Image](https://github.com/user-attachments/assets/918bd36d-a976-4fcb-95de-8bbd29d39e93)

**Comment by rusefillc** [comment]
@ElDominio @Abricosvw we now have https://github.com/rusefi/rusefi/issues/7150!

open question if just a curve would do, or do we need a proper 3D table?

**Comment by rusefillc** [comment]
@FDSoftware @dron0gus does anyone remember if this is now done?

**Comment by FDSoftware** [comment]
> [@FDSoftware](https://github.com/FDSoftware) [@dron0gus](https://github.com/dron0gus) does anyone remember if this is now done?

yes!, here: https://github.com/rusefi/rusefi/pull/7701

It would be good if someone tested it before closing it in an ideal world, but from what I could see from the simulator it is working as expected.

**Comment by Abricosvw** [comment]
it doesn't work for me... i followed the instructions as in this post... the west gate doesn't move at all ... I tried to select TPS, MAP does not work ...

**Comment by rusefillc** [comment]
@Abricosvw this ticket is too messy already, please create a new one with your tune attached?
